### DISCUSSION TITLE
standard i/o scriped "server" savers

### DISCUSSION THREAD

**Discussion by dubiouscript** [TiddlyWiki/TiddlyWiki5#6110] [General]
servers that use * "some program" to 
  listen on a port 
  and printing the incoming request(s) 
  which are read by the "server" script via standard input 
  that "responds" to request via standard output 

*netcat socat nc ( plus some arguments , script path , ect) ect

has any one come across any other examples of this kind of scriptable server ..

**Comment by dubiouscript** [comment]
shell script
- bashttpd 

 https://github.com/TooTallNate/bashttpd
 - bashttpd is a simple, configurable web server written in bash
- Simplify things by using pipes instead of fds https://github.com/TooTallNate/bashttpd/commit/e46ca8994eba703a2ffb853c39599f0e5c0ef9b0 :tada: 

https://gist.github.com/dubiouscript/e90a515b6ee2b5fe8988b945b0cdb418#file-bashttpd +gzip 

 https://gitlab.com/lordratte/tishlywiki ( written in bash and JavaScript and can probably run on most Linux environments ) 
 - server logic was hacked out of bashttpd 
 - inspired by Twexe
> run your TiddlyWiki file as an executable which starts a server and opens your TiddlyWiki file in your browser. Saving saves straight back to the file itself, making it an easy way to use a file on other devices without the need for plugins, etc.

... 
in adition to the above but on a similar theme

it should be posible to use jsfunc or similar 
https://github.com/TooTallNate/jsfunc - Write JavaScript functions - use as shell functions!

to convert some of the examples in 'Add support for running external shell tasks' PR
eg
https://github.com/Jermolene/TiddlyWiki5/pull/3461/files#diff-cbb8cb6f6e587b75dc16a66d77545fb31dcb68fb17fd1eb411d9e2af7bab0eb3

by removing the (repeated) server cfg 

>var net = require("net"),
	port = parseInt(process.argv[2] || "",10) || 8081, // Port
	host = process.argv[3] || "127.0.0.1"; // Host
  var server = net.createServer();
    server.listen(port,host);

and just reading/writing from i/o

**Comment by dubiouscript** [comment]
https://github.com/vi/websocat  Netcat, curl and socat for WebSockets.
 - Executing external program and making it communicate to WebSocket using stdin/stdout. 

https://github.com/joewalnes/websocketd
- turn any program that uses STDIN/STDOUT into a WebSocket server. Like inetd, but for WebSockets.

**Comment by dubiouscript** [comment]
:sunglasses: 

awk 

out of the few awk www servers i happen to have found 

this one ( crossbowerbt/awk-webserver ) 
https://github.com/crossbowerbt/awk-webserver
is perhaps the most succinct

or to put it in the parlance of modertimes :dog: 

awk-webserver wow!
finite  machine-statuses
most succinct

somuch so that while messing around "testing" it out 

i just copyed the raw request text from dev console into a varable 
echo'ed the requst text and piped that into 
eg
`  echo "$request_text" |  gawk -f ./webserver.awk ` 


the above is perhaps easyer to read 
than this 
another awk server example 
http://tuxgraphics.org/~guido/scripts/awk-one-liner.html
```
#!/usr/bin/gawk -f
BEGIN {
if (ARGC < 2) { print "Usage: wwwawk  file.html"; exit 0 }
	Concnt = 1;
        while (1) {
        RS = ORS = "\r\n";
        HttpService = "/inet/tcp/8080/0/0";
        getline Dat < ARGV[1];
        Datlen = length(Dat) + length(ORS);
        while (HttpService |& getline ){
		if (ERRNO) { print "Connection error: " ERRNO; exit 1}
                print "client: " $0;
                if ( length($0) < 1 ) break;
        }
        print "HTTP/1.1 200 OK"             |& HttpService;
        print "Content-Type: text/html"     |& HttpService;
        print "Server: wwwawk/1.0"          |& HttpService;
        print "Connection: close"           |& HttpService;
        print "Content-Length: " Datlen ORS |& HttpService;
        print Dat                           |& HttpService;
        close(HttpService);
        print "OK: served file " ARGV[1] ", count " Concnt;
        Concnt++;
      }
} 
```
*somefile.html eg empty.html 
```
chmod 755 wwwawk
./wwwawk somefile.html
```


which all though is shorter in LOC its less "composable" than the other example eg : handles the conection without need for socat ect 

that said 
i did consider that it might be intresting to reduce/modify 
the second awk svr example to 
work with the crossbowerbt/awk-webserver example (via io) replaceing the need for socat ect  :thinking: 
then i got distracted =/

... modifications to handle POST request/data paths 
might be useful / intresting additions to eather example 
which would allow for the posibility of 

-using existsing existing tw-savers , eg setting $:/UploadURL to point to the  "server"  http://address:port 

> https://github.com/Jermolene/TiddlyWiki5/issues/3946#issuecomment-494186655
> https://github.com/Jermolene/TiddlyWiki5/blob/2f8053265e96e0ae57b8d7b6c841af15aebd3f6d/core/modules/savers/upload.js#L60
> 
> ftr "0 - " means saved ;]

see also 
https://www.gnu.org/software/gawk/manual/gawkinet/gawkinet.html#Primitive-Service

:vulcan_salute:

**Comment by Jermolene** [comment]
Thanks @dubiouscript great stuff, some interesting tools there.

**Reply by dubiouscript** [reply]
@Jermolene  thanks ! 

i think crossbowerbt/awk-webserver :framed_picture:  is a *realy* a work of art , ( perhaps,.. modernist minimalism :rofl:  )

which has me thinking / reflecting on... ( amongst other things) 
the tag line "Have you ever had the feeling that your head is not quite big enough to hold everything you need to remember?"

a strange perspective is that i do remember ( some time ago )

reading through this *one page* of html ect 
https://pastebin.com/bs97y6qa - TiddlyWiki 1.0 (1400 lines )

and gocking it more or less and getting the *much wow* :dog: *very script* :exploding_head: 
(much fun/hacking was had/done)
and fitting enough of the 1400 lines in my  head to add more lines :scream: :wink: 

from my view it was composed of 
lots of bits of js "one liners" / concepts that  could be pull apart and  recombined
all in one page !
https://english.cdn.zeenews.com/sites/default/files/2021/06/15/944055-dog1.jpg

..as im writing the above and thinking of

0) "Rust, WebAssembly, and the future of Serverless" (also mind boggling) 

1) https://talk.tiddlywiki.org/t/is-tw-the-first-and-perhaps-only-epistemologically-and-ethically-correct-software-on-the-planet/12
...* i should probably register @talk.tiddlywiki.org but i lothe that much discourses functionality requires js (one day..)

2)https://tenfourfox.blogspot.com/2020/04/the-end-of-tenfourfox-and-what-ive.html

and more generally about the *shifting* space tiddlylywiki inhabits 
browsers have changed (arguably,.. into operating systems / containers )

which i think is perhaps in some ways a counter to the "epistemologically-and-ethically-correct-software"
argument or at least something else to consider while thinking about one peace of software in an *ever growing* ecosystem ( eg quote `jeanqasaur` : "A modern software application is a rainforest rather than a single tree." )

...[2] tenfourfox.blogspot provides some  example of the forces pacing/pushing modern javascript implementations 

and from those (i was going to say bazaar but they are realy more cathedral) forces 

more mind bogling
things like webasm emscripten , eg awk as "js" in browser 
```
strange/sureal that it is posible to surgest ..., however

https://github.com/petli-full/awkjs

https://awk.js.org/index.html
https://awk.js.org/assets/libs/mawk.js 
```

( some time ago ) when i was reading through early tiddlywiki 
learning various forms of "scripting"  , and escaping the broken windows bandwagon  
 
i remember looking in to the mysterious servers required to "host" the blooming web pages  
and finding (g)awk , evaluating it as looking interesting but lacking networking support on windrows 
and ended up finding *many* alternatives to host cgi scripts, and *eventually* booting/learning other systems 


Now upon finding crossbowerbt/awk-webserver it seems to bring me full circle or perhaps  full mobius strip :sweat_smile: 

but with a lot of interesting / unusual perspective from the journey =s 
(which makes me think of a venn-diagram 
between the 
conclusion of the  carlo rovelli "Physics and Philosophy of Time" lecture
and 
vinny jones closing remarks from lock stock and two smoking barrels 
which intersect at 
>"it's (been) emotional"
 ) 
/rant :v: 

---
just to note :sunglasses: 
some interesting 
pre processing : https://github.com/ttv1/aok/blob/master/docs/index.md
```
As an example of the last point,

emp.name.lname

expands to

emp["name"]["lname"]
```
and post processing : noyesno/awka ...
https://github.com/noyesno/awka/blob/master/docs/awka.svg

wrt to the latter as a starting point i imagine its not sooo :roll_eyes:  many steps away from  a wiki running embedded eg on microcontroller   :zany_face: 

in that regard it might be an interesting [s]excrement[/s]  :sweat: experiment  to try and distill / compress  some of the philosophical  essence ?
at least 
one can dream 
as they say :zzz: 
:

**Comment by dubiouscript** [comment]
another one , that just caught my attention 

not actually a server but related , but an example of using node in a pipe 
https://github.com/kevinoid/nodecat
